---
title: "03 Analyze Methylation"
author: "Julin Maloof"
date: "2023-11-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Goal

Find regions of differential methylation in L121 vs DaAe.  Using DaAe as reference.

Strategy:

probably using genome ranges.  Import the bed files.  Average across a region (window) and compare.

Make sure that there is coverage in both.

Possibly focus on genic or TSS regions only.

```{r}
library(tidyverse)
library(genomation)
library(GenomicRanges)
library(plyranges)
```

## import data

chromromes to keep (main chromosomes)
```{r}
chrkeep <- c(str_c("chrA", str_pad(1:10, 2, pad = "0")),
             str_c("chrC0", 1:9))
chrkeep
```


Get DaAe lengths and create GRanges object
```{r}
DaAeLengths.df <- read_delim("~/Sequences/ref_genomes/B_napus/John_DaAe/DaAe_20210520.fa.fai", col_names = c("chr", "length") ) %>%
  select(chr, length) %>% 
  filter(chr %in% chrkeep)

DaAeLengths.vec <- DaAeLengths.df %>% pull(length) %>% set_names(DaAeLengths.df$chr)

DaAeLengths.gr <-  GRanges(seqnames = str_c(DaAeLengths.df$chr, ":1-", DaAeLengths.df$length), seqlengths = DaAeLengths.vec)

DaAeLengths.gr
```

Get DaAe gff
```{r}
gff <- gffToGRanges("~/Sequences/ref_genomes/B_napus/John_DaAe/DaAe_20210520.gff3", filter = "mRNA")
gff <- gff[seqnames(gff) %in% chrkeep] 
seqlevels(gff) <- seqlevelsInUse(gff)

seqlengths(gff) <- DaAeLengths.vec
```

Create a TSS Granges.
```{r}
TSS <- promoters(gff, upstream=1000, downstream = 1000)
TSS
```



```{r}
L121 <- readGeneric("../input/cpg_scores/L121_to_DaAe_cpg.combined.bed",
                    meta.cols=list(L121_score=4,
                                   L121_coverage=6))

L121 <- L121[seqnames(L121) %in% chrkeep] 
seqlevels(L121) <- seqlevelsInUse(L121)
L121

L121
```

```{r}
DaAe <- readGeneric("../input/cpg_scores/DaAe_to_DaAe_cpg.combined.bed",
                    meta.cols=list(DaAe_score=4,
                                   DaAe_coverage=6))

DaAe <- DaAe[seqnames(DaAe) %in% chrkeep] 
seqlevels(DaAe) <- seqlevelsInUse(DaAe)
DaAe
```



## Intersect the files


```{r, warning=FALSE}
L121_DaAe <- join_overlap_intersect(L121, DaAe)
```

```{r}
L121_DaAe
```

```{r}
L121_DaAe %>% summarize(L121_score = mean(L121_score), DaAe_score = mean(DaAe_score))
``` 


```{r}

smL121 <- L121_DaAe %>% 
  mcolAsRleList("L121_score") %>%
  ScoreMatrix(windows = TSS, strand.aware = TRUE)

smDaAe <- L121_DaAe %>% 
  mcolAsRleList("DaAe_score") %>%
  ScoreMatrix(windows = TSS, strand.aware = TRUE)
```


```{r}
plotMeta(smL121, xcoords = -999:1000)
```

```{r}
plotMeta(smDaAe, xcoords = -999:1000)
```

A ggplot version of these plots, combined
```{r}
L121_DaAe.df <- as.data.frame(smDaAe) %>% 
  rownames_to_column(var = "gene") %>%
  pivot_longer(cols = -gene, names_to = "position", names_prefix = "V", values_to = "DaAe", values_drop_na = TRUE) 

L121_DaAe.df <- as.data.frame(smL121)  %>% 
  rownames_to_column(var = "gene") %>%
  pivot_longer(-gene, names_to = "position", names_prefix = "V", values_to = "L121", values_drop_na = TRUE) %>%
  full_join(L121_DaAe.df) %>%
  mutate(position = as.numeric(position) - 1000)

head(L121_DaAe.df, 1000)
```


```{r}
L121_DaAe.df %>%
  pivot_longer(c(L121, DaAe), values_to = "CpG", names_to = "ID") %>%
  ggplot(aes(x=position, y=CpG, color = ID)) +
  geom_smooth()
```



```{r fig.asp=1}
L121mean <- rowMeans(smL121[,500:1500], na.rm = TRUE)
DaAemean <- rowMeans(smDaAe[,500:1500], na.rm = TRUE)
means <- tibble(L121=L121mean, DaAe=DaAemean)
```

```{r, fig.asp=1}
means %>%
  ggplot(aes(x=DaAe, y=L121)) +
  geom_point(alpha=.2)
```

```{r}
means %>%
  summarize(higher_in_DaAe=sum(DaAe > 80 & L121 <= 20, na.rm = TRUE),
            higher_in_L121=sum(DaAe <= 20 & L121 > 80, na.rm = TRUE),
            total=dplyr::n(),
            total_with_data=sum(!is.na(DaAe)))
```

